src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,514 @@
+                500000,
+                1000000,
+                2000000,
+            };
+        }
+
+        /// <summary>
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        /// <exception cref="TimeoutException">There was no answer from the board</exception>
+        public virtual void Initialize()
+        {
+            lock (_initializationLock)
+            {
+                if (_initialized)
+                {
+                    return;
+                }
+
+                if (_isDisposed)
+                {
+                    throw new ObjectDisposedException("Board is already disposed");
+                }
+
+                if (_firmata != null)
+                {
+                    throw new InvalidOperationException("Board already initialized");
+                }
+
+                if (_serialPort != null)
+                {
+                    _serialPort.Open();
+                    _dataStream = _serialPort.BaseStream;
+                }
+                else if (_dataStream == null)
+                {
+                    // Should never get here
+                    throw new InvalidOperationException("Constructor argument error: Neither port nor stream specified");
+                }
+
+                _firmata = new FirmataDevice();
+                _firmata.Open(_dataStream);
+                _firmata.OnError += FirmataOnError;
+                _firmataVersion = _firmata.QueryFirmataVersion();
+                if (_firmataVersion < _firmata.QuerySupportedFirmataVersion())
+                {
+                    throw new NotSupportedException($"Firmata version on board is {_firmataVersion}. Expected {_firmata.QuerySupportedFirmataVersion()}. They must be equal.");
+                }
+
+                Log($"Firmata version on board is {_firmataVersion}.");
+
+                _firmwareVersion = _firmata.QueryFirmwareVersion(out var firmwareName);
+                _firmwareName = firmwareName;
+
+                Log($"Firmware version on board is {_firmwareVersion}");
+
+                _firmata.QueryCapabilities();
+
+                _supportedPinConfigurations = _firmata.PinConfigurations.AsReadOnly();
+
+                Log("Device capabilities: ");
+                foreach (var pin in _supportedPinConfigurations)
+                {
+                    Log(pin.ToString());
+                }
+
+                _firmata.EnableDigitalReporting();
+
+                _initialized = true;
+            }
+        }
+
+        /// <summary>
+        /// Firmware version on the device
+        /// </summary>
+        public Version FirmwareVersion
+        {
+            get
+            {
+                if (!_initialized)
+                {
+                    Initialize();
+                }
```

Comment:
see comment above, consider moving this logic into Initialize (currently it's there inside of the lock, but consider doing this check before and inside of the lock)